test/test_latest_time_policy.cpp

```diff
@@ -0,0 +1,331 @@
+    // or found to be late.
+    else if (idx >= 4U && idx < 6U)
+    {
+      EXPECT_EQ(h.count_, (idx + 2U) / 2U);
+      EXPECT_EQ(h.p_->data, p[3]->data);
+      EXPECT_EQ(h.q_->data, q[1]->data);
+      EXPECT_EQ(h.r_->data, r[0]->data);
+    }
+    // New actual rate of p computed when is received after q when idx==6
+    // for idx >= 6, follows normal "Leading" pattern again
+    // with pivot on q
+    else if (idx >= 6U)
+    {
+      EXPECT_EQ(h.count_, (idx + 2U) / 2U);
+      EXPECT_EQ(h.p_->data, p[idx / 2U]->data);
+      EXPECT_EQ(h.q_->data, q[idx / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[(idx - 2U) / 4U]->data);
+    }
+    else
+    {
+      EXPECT_FALSE(h.p_);
+      EXPECT_FALSE(h.q_);
+      EXPECT_FALSE(h.r_);
+    }
+
+    rate.sleep();
+  }
+
+}
+TEST_F(LatestTimePolicy, ChangeRateTrailing)
+{
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 12U; ++idx)
+  {
+    if(idx % 2U == 1U)
+    {
+      sync.add<1>(q[(idx - 1U) / 2U]);
+    }
+
+    if(idx % 4U == 3U)
+    {
+      sync.add<2>(r[(idx - 3U) / 4U]);
+    }
+
+    if (idx < 4U)
+    {
+      sync.add<0>(p[idx]);
+    }
+    else  // Change rate of p (still 1kHz @ idx == 4)
+    {
+      if(idx % 3U == 1U)
+      {
+        static std::size_t p_idx = 3U;
+        sync.add<0>(p[++p_idx]);
+      }
+    }
+
+    // operates like "Trailing" test for idx <= 3
+    if (idx > 2U && idx <= 4U)
+    {
+      EXPECT_EQ(h.count_, idx - 2U);
+      EXPECT_EQ(h.p_->data, p[idx]->data);
+      EXPECT_EQ(h.q_->data, q[(idx - 1U) / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[(idx - 3U) / 4U]->data);
+    }
+    // Rate of p still 1kHz @ idx==4.
+    // Then, change rate of p lower than q when idx==5.
+    // At idx==5, policy still doesn't know that p is late when q is received.
+    // Same behavior as initialization dropping faster messages until rates of all are known
+    // or found to be late.
+    else if (idx > 4U && idx < 7U)
+    {
+      EXPECT_EQ(h.count_, 2U);
+      EXPECT_EQ(h.p_->data, p[4U]->data);
+      EXPECT_EQ(h.q_->data, q[1U]->data);
+      EXPECT_EQ(h.r_->data, r[0U]->data);
+    }
+    // Will not publish again until idx==7, since rate of q is 500Hz
+    // and p is calculated as late when q recieved when idx==7 -- this makes q new pivot.
+    // Since q is new pivot and publishes when idx==7,
+    // and r comes in after q, r is now trailing.
+    // New actual rate of p computed when is received after q when idx==7
+    // for idx >= 7, follows normal "Trailing" pattern again
+    // with pivot on q
+    else if (idx >= 7U)
+    {
+      EXPECT_EQ(h.count_, (idx - 1U) / 2U);
+      EXPECT_EQ(h.p_->data, p[(idx + 1U) / 2U]->data);
+      EXPECT_EQ(h.q_->data, q[(idx - 1U) / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[(idx - 5U) / 4U]->data);
+    }
+    else
+    {
+      EXPECT_FALSE(h.p_);
+      EXPECT_FALSE(h.q_);
+      EXPECT_FALSE(h.r_);
+    }
+
+    rate.sleep();
```

Comment:
Instead of just sleeping a fixed amount of time between loops, is there a condition that could be checked here so it waits until that's satisfied? That would remove the reliance on time and make the test robust on slower machines.

One common pattern I've used is something like this:

```cpp
int sleep{0};
int maxSleep{30};
for (; condition && sleep < maxSleep; ++sleep)
{
  // sleep
}
EXPECT_LT(sleep, maxSleep);
EXPECT_TRUE(condition);
```

Reply:
Well, the rate sleep is for creating message rates.

I suppose instead of using a rate sleep, I could pass the `Policy` an `rclcpp::Clock` that is set for sim time, and control the `/clock` in the test? Then the time the messages are received would be deterministic.

Requiring stamped messages could also resolve timing issues. I was hoping to make it generic, but users should really be using stamps for messages that require timing anyways :stuck_out_tongue_winking_eye: